### ISSUE TITLE
Alltags view join to gps

### ISSUE THREAD

**Issue by cjardine-bsc** [MotusDev/Motus-TO-DO#421]
Phil uncovered a minor issue with the AllTags view, In looking at batch 34895 he noticed that the first few runs were missing GPS data, although the rest of the batch had that information. 

Digging into it, the allTags view is using a "<= ts" qualifier to decide what GPS fix a taghit belongs to. That means that if there is not a GPS fix before the taghit in the batch then the hits are not associated with a GPS point. 

LEFT JOIN
   sg_gps AS t11 ON t11.batchID = t3.batchID
      AND t11.ts =
         (SELECT
             max(t11b.ts)
          FROM
             sg_gps AS t11b
          WHERE
             t11b.batchID=t3.batchID
             AND t11b.ts <= t1.ts
         )

I would have expected a GPS fix to be taken at the beginning of every batch, but it seems like that's not always so. 

A fix would be to instead associate the taghit to the gps hit that is closest to it in time. This is probably better for users and mobile deployments anyway.  This doesn't seem to have a huge impact on performance.

LEFT JOIN
   sg_gps AS t11 ON t11.batchID = t3.batchID
      AND t11.ts =
         (select top 1 ts from sg_gps AS t11b where  t11b.batchID=t1.batchID order by ABS(t1.ts - t11b.ts) asc)

We should touch base and see if we need to tweak the alltags view.

**Comment by denislepage** [comment]
If we remain within the same batchID, I think the change below (above?) makes sense. A slightly more complicated option would be to extend this to any batches of the same receiver, but there would need to be an upper cap on the difference in ts. Ideally, the query would also report the ts of the gps fix used, but that would likely require an additional inner query.

**Comment by denislepage** [comment]
I wanted to post on other issues related to the alltags view, and was reminded about this pending issue. The above doesn't quite work in SQLite. Apparently you can't reference outer columns in the ORDER BY portion of a correlated subquery. A workaround would be like this (adapted to SQLite syntax):

LEFT JOIN
sg_gps AS t11 ON t11.batchID = t3.batchID
AND t11.ts =(select ts from (select ts, abs(ts - t1.ts) as diff_ts from gps where t11b.batchID=t3.batchID ) AS t11b order by diff_ts asc limit 1)

However, trials so far suggest that this has a very significant impact on the query speed. The project 1 database alltags query was slowed down by a factor of about 100x. This doesn't sound like a suitable approach.

**Comment by denislepage** [comment]
The other issue I wanted to post about was a discussion on the way alltags view resolves the deployments, so this is related. I experimented a bit with building a hitsDeployID table similar to what we use on the server. The idea would be to resolve deployments for tags and receivers once, and rely on those values in subsequent queries. I am not yet convinced this is a good approach, mainly because the data would need to be updated with any metadata update. On the project 1 database, the process takes about 80 seconds, which is about the same as running the alltags view in full. A modified version of the alltags view that relies on hitsDeployID take about 1/10 of the time (8 seconds).

Of course, a user would typically only load detections from a single tag a time, I would expect, which shortens considerably any individual calls to the database. The overhead of constantly updating the deployID values may not be worth the gain that happens on queries.

I also explored with indexing to see if I could improve the alltags query speed, but I haven't been able to make a significant dent.  For instance, adding an index on hits.runID could in theory make the join between runs and hits faster, but that has no appreciable impact. 

`CREATE INDEX `hits_runID` ON `hits` ( `runID` ASC )`

Here's the code I used to create and populate the table, as well as the revised code for alltags, if anyone wants to try to experiment. Might be worth trying to play some more with the largest project (not sure which one that is), but I suspect this is the sort of things that might better be done as a chapter for more advanced users on how they can manipulate the database tables, rather than a canned function. Or perhaps a post on the Google group.

```
-- ambiguous hits are duplicated, so the table needs motusTagID as a key
CREATE TABLE "hitsDeployID" ( `hitID` INTEGER NOT NULL, `motusTagID` INTEGER NOT NULL, `tagDeployID` INTEGER, `recvDeployID` INTEGER, PRIMARY KEY(`hitID`,`motusTagID`) )

-- insert records into hitsDeployID
delete from hitsDeployID
INSERT INTO hitsDeployID (hitID, motusTagID, tagDeployID, recvDeployID)
SELECT t1.hitID as hitID, CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end as motusTagID, t5.deployID as tagDeployID, t6.deployID as recvDeployID
FROM hits AS t1 LEFT JOIN runs AS t2 ON t1.runID = t2.runID 
left join allambigs t12 on t2.motusTagID = t12.ambigID 
LEFT JOIN batches AS t3 ON t3.batchID = t1.batchID 
LEFT JOIN tags AS t4 ON t4.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 
LEFT JOIN tagDeps AS t5 ON t5.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 
	AND t5.tsStart = (SELECT max(t5b.tsStart) FROM tagDeps AS t5b WHERE t5b.tagID = 
		CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end AND t5b.tsStart <= t1.ts AND (t5b.tsEnd IS NULL OR t5b.tsEnd >= t1.ts) ) 
LEFT JOIN recvs as t6a on t6a.deviceID =t3.motusDeviceID 
LEFT JOIN recvDeps AS t6 ON t6.deviceID = t3.motusDeviceID AND t6.tsStart = (SELECT max(t6b.tsStart) FROM recvDeps AS t6b WHERE t6b.deviceID=t3.motusDeviceID 
	AND t6b.tsStart <= t1.ts AND (t6b.tsEnd IS NULL OR t6b.tsEnd >= t1.ts) ) 

-- cached version of alltags
SELECT t1.hitID as hitID, t1.runID as runID, t1.batchID as batchID, t1.ts as ts, t1.sig as sig, t1.sigSD as sigsd, t1.noise as noise, t1.freq as freq, t1.freqSD as freqsd, t1.slop as slop, 
t1.burstSlop as burstSlop, t2.done as done, CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end as motusTagID, t12.ambigID as ambigID, t2.ant as port, t2.len as runLen, t3.monoBN as bootnum, 
t4.projectID as tagProjID, t4.mfgID as mfgID, t4.type as tagType, t4.codeSet as codeSet, t4.manufacturer as mfg, t4.model as tagModel, t4.lifeSpan as tagLifespan, t4.nomFreq as nomFreq, t4.bi as tagBI, 
t4.pulseLen as pulseLen, t5.deployID as tagDeployID, t5.speciesID as speciesID, t5.markerNumber as markerNumber, t5.markerType as markerType, t5.tsStart as tagDeployStart, t5.tsEnd as tagDeployEnd, 
t5.latitude as tagDeployLat, t5.longitude as tagDeployLon, t5.elevation as tagDeployAlt, t5.comments as tagDeployComments, ifnull(t5.fullID, printf('?proj?-%d#%s:%.1f', t5.projectID, t4.mfgID, t4.bi)) 
as fullID, t3.motusDeviceID as deviceID, t6.deployID as recvDeployID, t6.latitude as recvDeployLat, t6.longitude as recvDeployLon, t6.elevation as recvDeployAlt, t6a.serno as recv, 
t6.name as recvDeployName, t6.siteName as recvSiteName, t6.isMobile as isRecvMobile, t6.projectID as recvProjID, t7.antennaType as antType, t7.bearing as antBearing, t7.heightMeters as antHeight, 
t8.english as speciesEN, t8.french as speciesFR, t8.scientific as speciesSci, t8.`group` as speciesGroup, t9.label as tagProjName, t10.label as recvProjName, t11.lat as gpsLat, t11.lon as gpsLon, 
t11.alt as gpsAlt 
FROM hits AS t1 LEFT JOIN runs AS t2 ON t1.runID = t2.runID 
left join allambigs t12 on t2.motusTagID = t12.ambigID 
LEFT JOIN hitsDeployID t13 on t1.hitID = t13.hitID and CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end  = t13.motusTagID
LEFT JOIN batches AS t3 ON t3.batchID = t1.batchID 
LEFT JOIN tags AS t4 ON t4.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 
LEFT JOIN tagDeps AS t5 ON t5.deployID = t13.tagDeployID
LEFT JOIN recvs as t6a on t6a.deviceID =t3.motusDeviceID 
LEFT JOIN recvDeps AS t6 ON t6.deployID = t13.recvDeployID
LEFT JOIN antDeps AS t7 ON t7.deployID = t6.deployID AND t7.port = t2.ant 
LEFT JOIN species AS t8 ON t8.id = t5.speciesID 
LEFT JOIN projs AS t9 ON t9.ID = t5.projectID 
LEFT JOIN projs AS t10 ON t10.ID = t6.projectID 
LEFT JOIN gps AS t11 ON t11.batchID = t3.batchID AND t11.ts = (SELECT max(t11b.ts) FROM gps AS t11b WHERE t11b.batchID=t3.batchID AND t11b.ts <= t1.ts )
```

**Comment by denislepage** [comment]
For GPS, another approach to explore is do 2 queries, looking in each direction within the gps table. This one also relies on the cached deployID's. That's slower, but that one way to resolve the missing GPS detections in alltags.

```

SELECT t1.hitID as hitID, t1.runID as runID, t1.batchID as batchID, t1.ts as ts, t1.sig as sig, t1.sigSD as sigsd, t1.noise as noise, t1.freq as freq, t1.freqSD as freqsd, t1.slop as slop, 
t1.burstSlop as burstSlop, t2.done as done, CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end as motusTagID, t12.ambigID as ambigID, t2.ant as port, t2.len as runLen, t3.monoBN as bootnum, 
t4.projectID as tagProjID, t4.mfgID as mfgID, t4.type as tagType, t4.codeSet as codeSet, t4.manufacturer as mfg, t4.model as tagModel, t4.lifeSpan as tagLifespan, t4.nomFreq as nomFreq, t4.bi as tagBI, 
t4.pulseLen as pulseLen, t5.deployID as tagDeployID, t5.speciesID as speciesID, t5.markerNumber as markerNumber, t5.markerType as markerType, t5.tsStart as tagDeployStart, t5.tsEnd as tagDeployEnd, 
t5.latitude as tagDeployLat, t5.longitude as tagDeployLon, t5.elevation as tagDeployAlt, t5.comments as tagDeployComments, ifnull(t5.fullID, printf('?proj?-%d#%s:%.1f', t5.projectID, t4.mfgID, t4.bi)) 
as fullID, t3.motusDeviceID as deviceID, t6.deployID as recvDeployID, t6.latitude as recvDeployLat, t6.longitude as recvDeployLon, t6.elevation as recvDeployAlt, t6a.serno as recv, 
t6.name as recvDeployName, t6.siteName as recvSiteName, t6.isMobile as isRecvMobile, t6.projectID as recvProjID, t7.antennaType as antType, t7.bearing as antBearing, t7.heightMeters as antHeight, 
t8.english as speciesEN, t8.french as speciesFR, t8.scientific as speciesSci, t8.`group` as speciesGroup, t9.label as tagProjName, t10.label as recvProjName, ifnull(t11.lat,t14.lat) as gpsLat, ifnull(t11.lon,t14.lon) as gpsLon, 
ifnull(t11.alt,t14.alt) as gpsAlt 
FROM hits AS t1 LEFT JOIN runs AS t2 ON t1.runID = t2.runID 
left join allambigs t12 on t2.motusTagID = t12.ambigID 
LEFT JOIN hitsDeployID t13 on t1.hitID = t13.hitID and CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end  = t13.motusTagID
LEFT JOIN batches AS t3 ON t3.batchID = t1.batchID 
LEFT JOIN tags AS t4 ON t4.tagID = CASE WHEN t12.motusTagID is null then t2.motusTagID else t12.motusTagID end 
LEFT JOIN tagDeps AS t5 ON t5.deployID = t13.tagDeployID
LEFT JOIN recvs as t6a on t6a.deviceID =t3.motusDeviceID 
LEFT JOIN recvDeps AS t6 ON t6.deployID = t13.recvDeployID
LEFT JOIN antDeps AS t7 ON t7.deployID = t6.deployID AND t7.port = t2.ant 
LEFT JOIN species AS t8 ON t8.id = t5.speciesID 
LEFT JOIN projs AS t9 ON t9.ID = t5.projectID 
LEFT JOIN projs AS t10 ON t10.ID = t6.projectID 
LEFT JOIN gps AS t11 ON t11.batchID = t3.batchID AND t11.ts = (SELECT max(t11b.ts) FROM gps AS t11b WHERE t11b.batchID=t3.batchID AND t11b.ts <= t1.ts )
LEFT JOIN gps AS t14 ON t14.batchID = t3.batchID AND t14.ts = (SELECT max(t14b.ts) FROM gps AS t14b WHERE t14b.batchID=t3.batchID AND t14b.ts <= t1.ts )

```

**Comment by denislepage** [comment]
Reviving this. The alltags view is likely a source of frustration for many people given its VERY slow speed. I just did a test on a medium size database (17 Mb), and it ran for over 10 minutes. I've heard of cases exceeding an hour or never.

By far the biggest culprit is the lookup to the GPS table, based on limited testing.

`LEFT JOIN gps AS t11 ON t11.batchID = t3.batchID AND t11.ts = (SELECT max(t11b.ts) FROM gps AS t11b WHERE t11b.batchID = t3.batchID AND t11b.ts <= t1.ts )
`

For most receivers, which are static, this is not even useful. I think we should remove the t11 table from the view, and the 3 associated fields gpsLat, gpsLon and gpsAlt.

There may be a way to limit the query to only mobile deployments. That would leave some detections that aren't associated with a deployment without a GPS, but we'd have to provide instructions in the R book on how to go about checking this.

We could keep the existing view and rename it alltags_with_gps, or something like this, for convenience, though that will likely break some people's code.

**Comment by steffilazerte** [comment]
Summarizing from emails

**In the R package:**

- Rename current `alltags` view to `alltagsGPS`
- Remove gps columns from `alltags` view
- Help users add GPS values back into their data
  - Helper function
      - return gps index (only gps cols with `hitID`)
      - join by time bins (allow user to specify)
      - join by nearest ts
  - motus R book tutorial for by hand and with helper function

**Comment by adamdsmith** [comment]
Following up. I've not updated to the development version that renames `alltags`. However, I did use the `alltags_fast` view that @denislepage distributed. 

But I'm still getting terribly slow access to data. 

Here's an example on a pretty small query of a relatively small database (~21 MB). It can bog down for hours on our 500+ MB database.

```r
library(motus)
library(dplyr)

p96 <- tagme(96, update = FALSE)

# alltags and alltags_fast queries present?
all(c("alltags", "alltags_fast") %in% src_tbls(p96))
#> [1] TRUE

system.time({
  slow <- p96 %>% tbl("alltags") %>% 
    filter(recvSiteName == "Willow Point 2016") %>%
    select(tagProjID, speciesEN, mfgID, motusTagID:runLen, runID, ts, sig, noise, freqsd, 
           tagDeployID:tagDepLon, recvDeployID:recvDeployLon, recv, recvSiteName, 
           recvProjID, antType, antBearing)
  print(slow)
})
#>    user  system elapsed 
#>   10.27   23.31   33.58

system.time({
  fast <- p96 %>% tbl("alltags_fast") %>% 
    filter(recvSiteName == "Willow Point 2016") %>%
    select(tagProjID, speciesEN, mfgID, motusTagID:runLen, runID, ts, sig, noise, freqsd, 
           tagDeployID:tagDepLon, recvDeployID:recvDeployLon, recv, recvSiteName, 
           recvProjID, antType, antBearing)
  print(fast)
})
#>    user  system elapsed 
#>   10.03   23.08   33.11

system.time(slow_df <- collect(slow))
#>    user  system elapsed 
#>   65.46  145.01  210.50

system.time(fast_df <- collect(fast))
#>    user  system elapsed 
#>   66.01  140.64  206.65
```

**Comment by steffilazerte** [comment]
Would you be able to try this again with the beta3 version of the `motus` package and just access the `alltags` view (which should hopefully be faster now)?

```
remotes::install_github("MotusWTS/motus", ref = "beta3")
p96 <- tagme(96, update = FALSE)
fast <- dplyr::tbl(p96, "alltags")
slow <- dplyr::tbl(p96, "alltagsGPS")
```

That'll help me figure out if this is something specific to Denis' version or a problem with all of them.
Thanks!

**Comment by adamdsmith** [comment]
Sure thing. Stand by...

**Comment by adamdsmith** [comment]
First, `alltagsGPS` was not present, naturally. So when I run `p96 <- tagme(96, update = TRUE)` I get the following error:

```
updateMotusDb started (2 version update(s))
 - Move GPS data from alltags view to alltagsGPS view
 - Drop IS NOT NULL constraint on allambigs
Error in FUN(X[[i]], ...) : 
  Error : error in view alltags_fast: no such table: main.allambigs
```

Does this mean I need to completely re-download the project?

**Comment by steffilazerte** [comment]
Oh hmm, you can probably get away with deleting all the views? And then running the update:

```
p96 <- tagme(96, update = FALSE)
DBI::dbExecute(p96, "DROP VIEW IF EXISTS allambigs")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS alltagsGPS")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS alltags")
p96 <- tagme(96, update = TRUE)
```

**Comment by adamdsmith** [comment]
Currently updating. FYI, I had to do the following:

```r
p96 <- RSQLite::dbConnect(RSQLite::SQLite(), dbname = "project-96.motus")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS allambigs")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS alltagsGPS")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS alltags")
DBI::dbExecute(p96, "DROP VIEW IF EXISTS alltags_fast")
DBI::dbExecute(p96, "DROP TABLE IF EXISTS tagAmbig2")

p96 <- tagme(96, update = TRUE)
```

**Comment by adamdsmith** [comment]
Not a huge difference:

```r
system.time({
  fast <- dplyr::tbl(p96, "alltags") %>% collect()
})
#>   user  system elapsed 
#>  73.12  158.73  231.84 

system.time({
  slow <- dplyr::tbl(p96, "alltagsGPS") %>% collect()
})
#>   user  system elapsed 
#>  84.49  171.22  256.07 
```

**Comment by steffilazerte** [comment]
Hmm, good to know, I'll see if @denislepage has any insights

**Comment by denislepage** [comment]
Just a thought, but filtering on a non-indexed field can typically be quite slow in a database.


    filter(recvSiteName == "Willow Point 2016") %>%

Using a primary key such as a motusDeviceID, etc. could yield better results. Do you get similar time if you rather try to filter on a tagID?

The alltags view is simply a convenience to join multiple tables, but is definitely not the most efficient way to do this for all situations. A better approach would be to build your own joins among the underlying tables using dplyr, and only take the subset of data you need for each of them first. Each case will be unique, so it would be hard to come up with convenient tools to do this in all situations.

**Comment by cjardine-bsc** [comment]
This sparks something on my to-do list:  we should have an explanation of the SQL database structure, including a diagram, included in the R book so users can do this sort of thing. I'll work on drafting something. 

The only other thought I had would be to build a taghits_deploy_id table directly into R (I think it's determining the deployment based on ts that makes the alltags view so lethargic). A clean download could come with a completed taghits_deploy_id from our DB. But we'd need a function to keep it up to date with metadata updates. That function itself would be slow, but it would make the alltags view much speedier for subsequent calls.

**Comment by adamdsmith** [comment]
@denislepage, I tried a quick example filtering on `deviceID` vs. `recvSiteName`. Worlds faster. Which leads to @cjardine-bsc's thought---a diagram of the database structure/relationships would be extremely helpful. Thanks!

**Comment by denislepage** [comment]
Good. We should probably adjust the r book to clarify that the fields used for filtering may have a big impact.

**Comment by adamdsmith** [comment]
The benchmarking to quantify "worlds faster"...

``` r
library(motus)
library(dplyr)

p96 <- tagme(96, update = FALSE, dir = "~/FWS_Projects/REKN_NFWF/Data/")

system.time({
  use_pk <- p96 %>% tbl("alltags") %>% 
    # filter(recvSiteName == "Willow Point 2016") %>%
    filter(deviceID == 772) %>%
    select(tagProjID, speciesEN, mfgID, motusTagID:runLen, runID, ts, sig, noise, freqsd, 
           tagDeployID:tagDepLon, recvDeployID:recvDeployLon, recv, recvSiteName, 
           recvProjID, antType, antBearing)
  print(use_pk)
})
#>    user  system elapsed 
#>    0.21    0.06    0.26

system.time({
  no_use_pk <- p96 %>% tbl("alltags") %>% 
    filter(recvSiteName == "Willow Point 2016") %>%
    # filter(deviceID == 772) %>%
    select(tagProjID, speciesEN, mfgID, motusTagID:runLen, runID, ts, sig, noise, freqsd, 
           tagDeployID:tagDepLon, recvDeployID:recvDeployLon, recv, recvSiteName, 
           recvProjID, antType, antBearing)
  print(no_use_pk)
})
#>    user  system elapsed 
#>   10.45   23.52   33.96

system.time(use_pk <- collect(use_pk))
#>    user  system elapsed 
#>   11.23   24.53   35.76
system.time(no_use_pk <- collect(no_use_pk))
#>    user  system elapsed 
#>   69.15  158.12  228.17

identical(use_pk, no_use_pk)
#> [1] TRUE
```

**Comment by steffilazerte** [comment]
I'm assuming this isn't actually relevant to how motus creates the `alltags` and `alltagsGPS` views... right? I don't do any filtering anyway, unless you mean how the join depends on `ts` and `batchID`?

**Comment by adamdsmith** [comment]
@denislepage @cjardine-bsc  If you can get me something identifying the core tables and their PK and FK, I could try to hobble together a start at a workflow for custom joins using the new [`dm`](https://krlmlr.github.io/dm) package, which looks pretty interesting... Maybe start by trying to recreate `alltags` or similar...

I could probably track them down myself, but if you had them handy.

**Comment by denislepage** [comment]
Correct. There may be changes needed in the R book, but nothing in the R package for now.

**Comment by steffilazerte** [comment]
Hi everyone I'm back and would definitely like to get the beta3 branch merged with master soon. We just have a couple of outstanding things that need to be tested first. 

1. If I've understood correctly, although the new, faster `alltags` view isn't much much faster, it's better and addresses the issue here (@adamdsmith said it was faster if not worlds faster, @leberrigan can you confirm as well with beta3?)

```
src <- tagme(projectID, update=T)
tbl_alltags_fast <- tbl(src, "alltags") # New table without GPS (faster)
tbl_alltags_gps <- tbl(src, "alltagsGPS") # Old table with GPS (slower)
```

2. I'll update examples in motus R book that require GPS values so they use the "alltagsGPS" view. 

Once these two items are addressed, I think this issue is done (unless anyone has something else that needs to be addressed?)

**Comment by steffilazerte** [comment]
Motus R book is updated on beta branch

**Comment by steffilazerte** [comment]
I think this issue has been resolved, if the people want to explore faster, custom joins using the new dm package I suggest we open a new issue for that specifically.
